{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create conversion maps from human genome files\n",
    "\n",
    "Human gene info can be downloaded from ftp://ftp.ncbi.nih.gov/gene/DATA/GENE_INFO/Mammalia/Homo_sapiens.gene_info.gz and gunzipped to a text file. This text file can be parsed with this notebook to create python dict objects used to map any entrez_id to a gene_symbol or vice-versa.\n",
    "\n",
    "Some of these functions are copied from PyGEST, and cleaned up for more general use."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" Change this to True to download and unpack a fresh genome to the local directory. \"\"\"\n",
    "download_new_file = False\n",
    "# human_genome_file = '/data/sourcedata/Homo_sapiens.gene_info'\n",
    "human_genome_file = './Homo_sapiens.gene_info'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" Download a fresh version of the human genome annotation, if desired. \"\"\"\n",
    "if download_new_file:\n",
    "    import urllib.request as request\n",
    "    import gzip\n",
    "    \n",
    "    baseURL = \"ftp://ftp.ncbi.nih.gov/gene/DATA/GENE_INFO/Mammalia\"\n",
    "    filename = \"Homo_sapiens.gene_info.gz\"\n",
    "\n",
    "    response = request.urlopen(\"/\".join([baseURL, filename, ]))\n",
    "    with open(human_genome_file, \"wb\") as f:\n",
    "        f.write(gzip.decompress(response.read()))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" For this notebook, load the gene_info file and keep it in memory. \"\"\"\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "gene_info_file = human_genome_file\n",
    "\n",
    "human_genome_info = pd.read_csv(gene_info_file, delimiter='\\t')\n",
    "human_genome_info = human_genome_info.set_index('GeneID')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#tax_id</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>LocusTag</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>dbXrefs</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>map_location</th>\n",
       "      <th>description</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>Symbol_from_nomenclature_authority</th>\n",
       "      <th>Full_name_from_nomenclature_authority</th>\n",
       "      <th>Nomenclature_status</th>\n",
       "      <th>Other_designations</th>\n",
       "      <th>Modification_date</th>\n",
       "      <th>Feature_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GeneID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>-</td>\n",
       "      <td>A1B|ABG|GAB|HYST2477</td>\n",
       "      <td>MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410</td>\n",
       "      <td>19</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>O</td>\n",
       "      <td>alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...</td>\n",
       "      <td>20191012</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606</td>\n",
       "      <td>A2M</td>\n",
       "      <td>-</td>\n",
       "      <td>A2MD|CPAMD5|FWP007|S863-7</td>\n",
       "      <td>MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>O</td>\n",
       "      <td>alpha-2-macroglobulin|C3 and PZP-like alpha-2-...</td>\n",
       "      <td>20191012</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>-</td>\n",
       "      <td>A2MP</td>\n",
       "      <td>HGNC:HGNC:8|Ensembl:ENSG00000256069</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>O</td>\n",
       "      <td>pregnancy-zone protein pseudogene</td>\n",
       "      <td>20191012</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9606</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>-</td>\n",
       "      <td>AAC1|MNAT|NAT-1|NATI</td>\n",
       "      <td>MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171428</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>O</td>\n",
       "      <td>arylamine N-acetyltransferase 1|N-acetyltransf...</td>\n",
       "      <td>20191103</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>9606</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>-</td>\n",
       "      <td>AAC2|NAT-2|PNAT</td>\n",
       "      <td>MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156006</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>O</td>\n",
       "      <td>arylamine N-acetyltransferase 2|N-acetyltransf...</td>\n",
       "      <td>20191014</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8923215</th>\n",
       "      <td>741158</td>\n",
       "      <td>trnD</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MT</td>\n",
       "      <td>-</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>20180406</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8923216</th>\n",
       "      <td>741158</td>\n",
       "      <td>trnP</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MT</td>\n",
       "      <td>-</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>20180406</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8923217</th>\n",
       "      <td>741158</td>\n",
       "      <td>trnA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MT</td>\n",
       "      <td>-</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>20180406</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8923218</th>\n",
       "      <td>741158</td>\n",
       "      <td>COX1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MT</td>\n",
       "      <td>-</td>\n",
       "      <td>cytochrome c oxidase subunit I</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>cytochrome c oxidase subunit I</td>\n",
       "      <td>20180406</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8923219</th>\n",
       "      <td>741158</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MT</td>\n",
       "      <td>-</td>\n",
       "      <td>l-rRNA</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>20180406</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61673 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         #tax_id    Symbol LocusTag                   Synonyms  \\\n",
       "GeneID                                                           \n",
       "1           9606      A1BG        -       A1B|ABG|GAB|HYST2477   \n",
       "2           9606       A2M        -  A2MD|CPAMD5|FWP007|S863-7   \n",
       "3           9606     A2MP1        -                       A2MP   \n",
       "9           9606      NAT1        -       AAC1|MNAT|NAT-1|NATI   \n",
       "10          9606      NAT2        -            AAC2|NAT-2|PNAT   \n",
       "...          ...       ...      ...                        ...   \n",
       "8923215   741158      trnD        -                          -   \n",
       "8923216   741158      trnP        -                          -   \n",
       "8923217   741158      trnA        -                          -   \n",
       "8923218   741158      COX1        -                          -   \n",
       "8923219   741158  16S rRNA        -                          -   \n",
       "\n",
       "                                                   dbXrefs chromosome  \\\n",
       "GeneID                                                                  \n",
       "1           MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410         19   \n",
       "2           MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899         12   \n",
       "3                      HGNC:HGNC:8|Ensembl:ENSG00000256069         12   \n",
       "9        MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171428          8   \n",
       "10       MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156006          8   \n",
       "...                                                    ...        ...   \n",
       "8923215                                                  -         MT   \n",
       "8923216                                                  -         MT   \n",
       "8923217                                                  -         MT   \n",
       "8923218                                                  -         MT   \n",
       "8923219                                                  -         MT   \n",
       "\n",
       "        map_location                         description    type_of_gene  \\\n",
       "GeneID                                                                     \n",
       "1           19q13.43              alpha-1-B glycoprotein  protein-coding   \n",
       "2           12p13.31               alpha-2-macroglobulin  protein-coding   \n",
       "3           12p13.31  alpha-2-macroglobulin pseudogene 1          pseudo   \n",
       "9               8p22               N-acetyltransferase 1  protein-coding   \n",
       "10              8p22               N-acetyltransferase 2  protein-coding   \n",
       "...              ...                                 ...             ...   \n",
       "8923215            -                                tRNA            tRNA   \n",
       "8923216            -                                tRNA            tRNA   \n",
       "8923217            -                                tRNA            tRNA   \n",
       "8923218            -      cytochrome c oxidase subunit I  protein-coding   \n",
       "8923219            -                              l-rRNA            rRNA   \n",
       "\n",
       "        Symbol_from_nomenclature_authority  \\\n",
       "GeneID                                       \n",
       "1                                     A1BG   \n",
       "2                                      A2M   \n",
       "3                                    A2MP1   \n",
       "9                                     NAT1   \n",
       "10                                    NAT2   \n",
       "...                                    ...   \n",
       "8923215                                  -   \n",
       "8923216                                  -   \n",
       "8923217                                  -   \n",
       "8923218                                  -   \n",
       "8923219                                  -   \n",
       "\n",
       "        Full_name_from_nomenclature_authority Nomenclature_status  \\\n",
       "GeneID                                                              \n",
       "1                      alpha-1-B glycoprotein                   O   \n",
       "2                       alpha-2-macroglobulin                   O   \n",
       "3          alpha-2-macroglobulin pseudogene 1                   O   \n",
       "9                       N-acetyltransferase 1                   O   \n",
       "10                      N-acetyltransferase 2                   O   \n",
       "...                                       ...                 ...   \n",
       "8923215                                     -                   -   \n",
       "8923216                                     -                   -   \n",
       "8923217                                     -                   -   \n",
       "8923218                                     -                   -   \n",
       "8923219                                     -                   -   \n",
       "\n",
       "                                        Other_designations  Modification_date  \\\n",
       "GeneID                                                                          \n",
       "1        alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...           20191012   \n",
       "2        alpha-2-macroglobulin|C3 and PZP-like alpha-2-...           20191012   \n",
       "3                        pregnancy-zone protein pseudogene           20191012   \n",
       "9        arylamine N-acetyltransferase 1|N-acetyltransf...           20191103   \n",
       "10       arylamine N-acetyltransferase 2|N-acetyltransf...           20191014   \n",
       "...                                                    ...                ...   \n",
       "8923215                                                  -           20180406   \n",
       "8923216                                                  -           20180406   \n",
       "8923217                                                  -           20180406   \n",
       "8923218                     cytochrome c oxidase subunit I           20180406   \n",
       "8923219                                                  -           20180406   \n",
       "\n",
       "        Feature_type  \n",
       "GeneID                \n",
       "1                  -  \n",
       "2                  -  \n",
       "3                  -  \n",
       "9                  -  \n",
       "10                 -  \n",
       "...              ...  \n",
       "8923215            -  \n",
       "8923216            -  \n",
       "8923217            -  \n",
       "8923218            -  \n",
       "8923219            -  \n",
       "\n",
       "[61673 rows x 15 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_genome_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def id_to_symbol_map():\n",
    "    return human_genome_info['Symbol'].to_dict()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def symbol_to_id_map(use_synonyms=True, print_dupes=False):\n",
    "    \"\"\"\n",
    "    Load gene info file and convert it to a dictionary allowing rapid entrez_id lookup from symbols\n",
    "\n",
    "    :param use_synonyms: Set to False to only use gene symbols from the symbol column. By default, synonyms match too.\n",
    "    :param print_dupes: Set to True to print out each time an entrez_id is overwritten during map creation.\n",
    "    :return: dictionary mapping symbols to entrez ids\n",
    "    \"\"\"\n",
    "\n",
    "    syn_map = {}\n",
    "    sid_map = {}\n",
    "    symbols = set()\n",
    "    \n",
    "    # Map synonyms first. They can later be overwritten by primary symbols\n",
    "    if use_synonyms:\n",
    "        for i, row in human_genome_info.sort_index(ascending=False).iterrows():\n",
    "            for symbol in row['Synonyms'].split(\"|\"):\n",
    "                symbols.add(symbol)\n",
    "                # Store each synonymous symbol string as a key in the dictionary,\n",
    "                if symbol in syn_map.keys() and row.name != syn_map[symbol]:\n",
    "                    if print_dupes:\n",
    "                        print(\"  appending synonym '{}' to {{{}:{}}}\".format(row.name, symbol, syn_map[symbol]))\n",
    "                    if isinstance(syn_map[symbol], list):\n",
    "                        syn_map[symbol].append(row.name)\n",
    "                    else:\n",
    "                        syn_map[symbol] = [syn_map[symbol], row.name]\n",
    "                else:\n",
    "                    syn_map[symbol] = row.name\n",
    "\n",
    "    # Reverse-map Entrez IDs and Symbols\n",
    "    for i, row in human_genome_info.sort_index(ascending=False).iterrows():\n",
    "        symbols.add(row['Symbol'])\n",
    "        if row['Symbol'] in sid_map.keys() and row.name != sid_map[row['Symbol']]:\n",
    "            if print_dupes:\n",
    "                print(\"  appending id '{}' to {{{}:{}}}\".format(row.name, row['Symbol'], sid_map[row['Symbol']]))\n",
    "            if isinstance(sid_map[row['Symbol']], list):\n",
    "                sid_map[row['Symbol']].append(row.name)\n",
    "            else:\n",
    "                sid_map[row['Symbol']] = [sid_map[row['Symbol']], row.name, ]\n",
    "        # Store the canonical symbol string as a key in the dictionary\n",
    "        else:\n",
    "            sid_map[row['Symbol']] = row.name\n",
    "\n",
    "    # Remove the empty symbol\n",
    "    try:\n",
    "        symbols.remove(\"-\")\n",
    "    except KeyError:\n",
    "        pass\n",
    "\n",
    "    # Determine the appropriate entrez_id to use for each gene symbol.\n",
    "    symbol_list = []\n",
    "    for gene in sorted(list(symbols)):\n",
    "        this_gene = {'gene': gene, 'syn_hits': 0, 'id_hits': 0}\n",
    "        # Try synonyms first, then they can be overwritten if necessary.\n",
    "        try:\n",
    "            if isinstance(syn_map[gene], list):\n",
    "                this_gene['entrez_id'] = 0\n",
    "                this_gene['syn_id'] = None\n",
    "                this_gene['syn_hits'] = len(syn_map[gene])\n",
    "            else:\n",
    "                this_gene['entrez_id'] = int(syn_map[gene])\n",
    "                this_gene['syn_id'] = int(syn_map[gene])\n",
    "                this_gene['syn_hits'] = 1\n",
    "        except KeyError:\n",
    "            pass\n",
    "        # IDs are priority. If we find one, overwrite a synonym.\n",
    "        try:\n",
    "            if isinstance(sid_map[gene], list):\n",
    "                this_gene['entrez_id'] = 0\n",
    "                this_gene['sid_id'] = None\n",
    "                this_gene['id_hits'] = len(sid_map[gene])\n",
    "            else:\n",
    "                this_gene['entrez_id'] = int(sid_map[gene])\n",
    "                this_gene['sid_id'] = int(sid_map[gene])\n",
    "                this_gene['id_hits'] = 1\n",
    "        except KeyError:\n",
    "            # No id, there may already be a synonym. If so, leave it alone.\n",
    "            pass\n",
    "        symbol_list.append(this_gene)\n",
    "\n",
    "    # Manually add a few that are in AHBA, but not in the NCBI file.\n",
    "    symbol_list.append({'gene': 'FLJ23867', 'entrez_id': 200058})\n",
    "    symbol_list.append({'gene': 'FLJ37035', 'entrez_id': 399821})\n",
    "    symbol_list.append({'gene': 'FLJ21408', 'entrez_id': 400512})\n",
    "    symbol_list.append({'gene': 'PP14571', 'entrez_id': 100130449})\n",
    "\n",
    "    # Convert to a dataframe, report stats, and update the dictionary.\n",
    "    df_symbols = pd.DataFrame(data=symbol_list)\n",
    "    if use_synonyms:\n",
    "        print(\"Individually, {} synonyms, {} ids\".format(len(syn_map), len(sid_map)))\n",
    "        print(\"Combined, {} synonyms, {} ids\".format(\n",
    "            len(df_symbols[df_symbols['syn_id'].notnull()]), len(df_symbols[df_symbols['sid_id'].notnull()])\n",
    "        ))\n",
    "    else:\n",
    "        print(\"Individually, {} ids\".format(len(sid_map)))\n",
    "        print(\"Combined, {} ids\".format(len(df_symbols[df_symbols['sid_id'].notnull()])))\n",
    "    print(\"{} good keys. {} have ambiguous (multiple) mappings, {} nulls\".format(\n",
    "        len(df_symbols[df_symbols['entrez_id'] > 0]),\n",
    "        len(df_symbols[df_symbols['entrez_id'] == 0]),\n",
    "        len(df_symbols[df_symbols['entrez_id'].isnull()]),\n",
    "    ))\n",
    "\n",
    "    return df_symbols.set_index('gene')['entrez_id'].to_dict()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" Create an id to symbol map and save it to disk. \"\"\"\n",
    "\n",
    "import pickle\n",
    "\n",
    "i_s = id_to_symbol_map()\n",
    "with open(\"./id_to_symbol_map.dict\", \"wb\") as f:\n",
    "    pickle.dump(i_s, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Individually, 64297 synonyms, 61287 ids\n",
      "Combined, 61056 synonyms, 61236 ids\n",
      "120953 good keys. 3101 have ambiguous (multiple) mappings, 0 nulls\n"
     ]
    }
   ],
   "source": [
    "\"\"\" Create a symbol to id map and save it to disk. \"\"\"\n",
    "\n",
    "import pickle\n",
    "\n",
    "s_i = symbol_to_id_map()\n",
    "with open(\"./symbol_to_id_map.dict\", \"wb\") as f:\n",
    "    pickle.dump(s_i, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "schmidt_geconn",
   "language": "python",
   "name": "schmidt_geconn"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
